gitlab/resource_gitlab_repository_file.go

```diff
@@ -0,0 +1,212 @@
+)
+
+const encoding = "base64"
+
+func resourceGitLabRepositoryFile() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGitlabRepositoryFileCreate,
+		Read:   resourceGitlabRepositoryFileRead,
+		Update: resourceGitlabRepositoryFileUpdate,
+		Delete: resourceGitlabRepositoryFileDelete,
+		Importer: &schema.ResourceImporter{
+			State: func(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				s := strings.Split(d.Id(), ":")
+
+				if len(s) != 3 {
+					d.SetId("")
+					return nil, fmt.Errorf("invalid Repository File import format; expected '{project_id}:{branch}:{file_path}'")
+				}
+				project, branch, filePath := s[0], s[1], s[2]
+
+				d.SetId(filePath)
+				d.Set("project", project)
+				d.Set("branch", branch)
+
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		// the schema matches https://docs.gitlab.com/ee/api/repository_files.html#create-new-file-in-repository
+		// However, we don't support the `encoding` parameter as it seems to be broken.
+		// Only a value of `base64` is supported, all others, including the documented default `text`, lead to
+		// a `400 {error: encoding does not have a valid value}` error.
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"file_path": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"branch": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"start_branch": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"author_email": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"author_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"content": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validateBase64Content,
+			},
+			"commit_message": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"encoding": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabRepositoryFileCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(string)
+	filePath := d.Get("file_path").(string)
+
+	options := &gitlab.CreateFileOptions{
+		Branch:        gitlab.String(d.Get("branch").(string)),
+		Encoding:      gitlab.String(encoding),
+		AuthorEmail:   gitlab.String(d.Get("author_email").(string)),
+		AuthorName:    gitlab.String(d.Get("author_name").(string)),
+		Content:       gitlab.String(d.Get("content").(string)),
+		CommitMessage: gitlab.String(d.Get("commit_message").(string)),
+	}
+	if startBranch, ok := d.GetOk("start_branch"); ok {
+		options.StartBranch = gitlab.String(startBranch.(string))
+	}
+
+	repositoryFile, _, err := client.RepositoryFiles.CreateFile(project, filePath, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(repositoryFile.FilePath)
```

Comment:
If the user has two gitlab_repository_file resources for the same file but different projects or branches, wouldn't the ID cause a state conflict? I think we should use the ID format that you have in the importer.

Reply:
Yes, good catch. I've just updated the implementation using `project:branch:filePath` as id and also using the `StatePassThroughContext` importer instead of a custom one.

Reply:
Perfect